corehq/apps/userreports/tests/test_json_query_expressions.py

```diff
@@ -0,0 +1,455 @@
+        []
+    ),
+    (
+        ["a", "b"],
+        {'type': 'identity'},
+        ["a", "b"],
+    ),
+    (
+        [1, 2],
+        {'type': 'identity'},
+        [],
+    ),
+])
+def test_flatten_basic(self, doc, items_ex, expected):
+    expression = ExpressionFactory.from_spec({
+        'type': 'flatten',
+        'items_expression': items_ex
+    })
+    self.assertEqual(expression(doc), expected)
+
+
+@generate_cases([
+    ([{'key': 'v1'}, {'key': 'v2'}], None),
+    ([{'key': 'v1'}, {'key': 'v2'}], {}),
+    (None, {'type': 'identity'}),
+    ({}, {'type': 'identity'}),
+])
+def test_sort_items_bad_spec(self, items_ex, sort_ex):
+    expression = {
+        'type': 'sort_items',
+    }
+    if items_ex is not None:
+        expression['items_expression'] = items_ex
+    if sort_ex is not None:
+        expression['sort_expression'] = sort_ex
+
+    with self.assertRaises(BadSpecError):
+        ExpressionFactory.from_spec(expression)(items_ex, sort_ex)
+
+
+@generate_cases([
+    (
+        {'items': [{'key': 2}, {'key': 1}]},
+        {'type': 'property_name', 'property_name': 'items'},
+        {'type': 'property_name', 'property_name': 'key'},
+        [{'key': 1}, {'key': 2}],
+    ),
+    (
+        [{'key': 2}, {'key': 1}],
+        {'type': 'identity'},
+        {'type': 'property_name', 'property_name': 'key'},
+        [{'key': 1}, {'key': 2}],
+    ),
+    (
+        {'items': [2, 1, 3]},
+        {'type': 'property_name', 'property_name': 'items'},
+        {'type': 'identity'},
+        [1, 2, 3],
+    ),
+    (
+        [2, 1, 3],
+        {'type': 'identity'},
+        {'type': 'identity'},
+        [1, 2, 3],
+    ),
+    # any possible python object comparison is supported
+    (
+        [datetime.date(2013, 1, 2), datetime.date(2013, 1, 1), datetime.date(2013, 1, 3)],
+        {'type': 'identity'},
+        {'type': 'identity'},
+        [datetime.date(2013, 1, 1), datetime.date(2013, 1, 2), datetime.date(2013, 1, 3)],
+    ),
+    # if sort_by returns empty, items are left in the same order
+    (
+        {'items': [{'key': 2}, {'key': 1}]},
+        {'type': 'property_name', 'property_name': 'items'},
+        {'type': 'property_name', 'property_name': 'no_match'},
+        [{'key': 2}, {'key': 1}],  # no sort
+    ),
+    # if sort_by returns empty for an item, it's inserted at the beginning of list
+    (
+        {'items': [{'key': 2}, {'no_key_match': 3}, {'key': 1}]},
+        {'type': 'property_name', 'property_name': 'items'},
+        {'type': 'property_name', 'property_name': 'key'},
+        [{'no_key_match': 3}, {'key': 1}, {'key': 2}],   # no sort
+    ),
+    # if items_expression doesn't match, empty list is returned
+    (
+        {'items': [{'key': 2}, {'key': 1}]},
+        {'type': 'property_name', 'property_name': 'no_match'},
+        {'type': 'property_name', 'property_name': 'key'},
+        [],
+    ),
+    # if sort values of two items can't be compared, empty list is returned
```

Comment:
just musing - this is a pretty weird failure mode but i'm not sure i have a better proposal